Update docstrings to numpy style, add it to linting checks.

bhazelton opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Updated docstrings throughout. This builds on work started by @aelanman. 

Also upgraded the RTD yaml to the newest version.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. If this PR closes an issue, put the word 'closes' before the issue link to auto-close the issue when the PR is merged. -->
closes #230 

## Types of changes
<!--- What types of changes does your code introduce? Put an replace the space with an `x` in all the boxes that apply: -->
- [x] Documentation change (documentation changes only)


## Checklist:
<!--- You may remove the checklists that don't apply to your change type(s) or just leave them empty -->
<!--- Go over all the following points, and replace the space with an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
For all pull requests:
- [x] I have read the [contribution guide](CONTRIBUTING.md).
- [x] My code follows the code style of this project.

Documentation change checklist:
- [x] Any updated docstrings use the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [x] If this is a significant change to the readme or other docs, I have checked that they are rendered properly on ReadTheDocs. (you may need help to get this branch to build on RTD, just ask!)

steven-murray commented on pyuvsim/baseline.py:
```diff
@@ -1,41 +1,66 @@
 # -*- mode: python; coding: utf-8 -*
 # Copyright (c) 2018 Radio Astronomy Software Group
 # Licensed under the 3-clause BSD License
+"""Definition of Baseline objects, to describe a pair of Antennas."""
```
autoref Antenna

steven-murray commented on pyuvsim/baseline.py:
```diff
@@ -1,41 +1,66 @@
     """
-    Defines a single pair of Antenna objects, the means for calculating their
-    uvw coordinates, and comparing them
+    Defines a single pair of Antenna objects with useful methods.
```
autoref antenna

steven-murray commented on pyuvsim/analyticbeam.py:
```diff
@@ -14,18 +15,25 @@
-               with the given diameter.
+    Parameters
+    ----------
+    diam: float
```
colon needs a space

steven-murray commented on pyuvsim/mpi.py:
```diff
@@ -343,12 +354,20 @@ class Counter:
 
+    Parameters
+    ----------
+    comm : TODO: Fill this out.
```
flagging a TODO

steven-murray commented on pyuvsim/mpi.py:
```diff
@@ -366,9 +385,18 @@ def __init__(self, comm=None, count_rank=0):
 
     def next(self, increment=1):
+        """
+        TODO: Fill this out.
```
flagging a TODO

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -33,15 +42,27 @@
 
+    Columns in the file provide, in order from left to right, the antenna name, antenna number,
+    a beam ID number, and the antenna positions relative to the telescope location in
+    east, north, up (ENU) in meters. See docs/parameter_files.rst for more details.
```
I wonder if this wouldn't be better to just have a URL pointing to RTD

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -126,7 +176,38 @@ def _set_lsts_on_uvdata(uv_obj):
     map_nside, diffuse_model, diffuse_params, time, localframe, array_location
 ):
+    """
+    Make a SkyModel object from an analog diffuse map function.
```
autoref skymdel

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -126,7 +176,38 @@ def _set_lsts_on_uvdata(uv_obj):
+        Name of diffuse model.
+    diffuse_params : dict
+        Dictionary of parameters corresponding to the `diffuse_model`.
+    time : astropy Time object
```
autoref `astropy.Time` -- you can add astropy to the `intersphinx_mapping` in the `docs/conf.py` and it will get the relevant link

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -126,7 +176,38 @@ def _set_lsts_on_uvdata(uv_obj):
+    -------
+    catalog : class:`pyradiosky.SkyModel`
+        SkyModel object containing the diffuse map.
+    icrs_coord : astropy SkyCoord
```
autoref

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -170,6 +251,36 @@ def _create_catalog_diffuse(
 
 def _create_catalog_discrete(Nsrcs, alts, azs, fluxes, time, localframe, array_location):
+    """
+    Make a SkyModel object from a set of point sources.
```
autoref

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -170,6 +251,36 @@ def _create_catalog_diffuse(
+        Azimuths or Longitues (depending on localframe) of the sources in radians.
+    fluxes : array_like of float
+        Brightnesses of the sources in Jy.
+    time : astropy Time object
```
autoref

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -170,6 +251,36 @@ def _create_catalog_diffuse(
+        Brightnesses of the sources in Jy.
+    time : astropy Time object
+        Time to use in defining the positions.
+    localframe : astropy BaseCoordinateFrame class or str
```
autoref

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -170,6 +251,36 @@ def _create_catalog_diffuse(
+        Time to use in defining the positions.
+    localframe : astropy BaseCoordinateFrame class or str
+        Frame that the source positions are defined in (e.g. altaz or ICRS)
+    array_location : astropy EarthLocation or MoonLocation
```
autoref

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -170,6 +251,36 @@ def _create_catalog_diffuse(
+    -------
+    catalog : class:`pyradiosky.SkyModel`
+        SkyModel object containing the diffuse map.
+    icrs_coord : astropy SkyCoord
```
autoref

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -1057,18 +1193,24 @@ def parse_frequency_params(freq_params):
-            https://pyuvsim.readthedocs.io/en/latest/parameter_files.html#frequency
+    Parameters
+    ----------
+    freq_params: dict
```
space before colon

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -1161,17 +1303,23 @@ def parse_time_params(time_params):
-            https://pyuvsim.readthedocs.io/en/latest/parameter_files.html#time
+    Parameters
+    ----------
+    time_params: dict
```
space before colon

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -1269,14 +1417,24 @@ def parse_time_params(time_params):
-    to a given frequency array.
+    Get a set of parameters that can be used to generate a given frequency array.
+
+    Returns a dictionary of parameters that can be used by parse_freq_params
```
autoref `parse_freq_params`

bhazelton replied:
Actually it seems that there is no longer any such function. I'm just removing this sentence.

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -1762,32 +1936,39 @@ def uvdata_to_telescope_config(
-        if return_names, returns tuple (path, telescope_config_name, layout_csv_name)
+    Parameters
+    ----------
+    uvdata_in: pyuvdata.UVData
```
colons here

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -1833,17 +2014,27 @@ def uvdata_to_config_file(uvdata_in, param_filename=None, telescope_config_name=
 
-    Args:
-        uvdata_in (UVData): uvdata object.
+    When used with `uvdata_to_telescope_config`, this will produce all the necessary
```
autoref

steven-murray commented on pyuvsim/simsetup.py:
```diff
@@ -1833,17 +2014,27 @@ def uvdata_to_config_file(uvdata_in, param_filename=None, telescope_config_name=
-    """
+    Parameters
+    ----------
+    uvdata_in: pyuvdata.UVData
```
autoref

steven-murray commented on pyuvsim/telescope.py:
```diff
@@ -387,3 +354,51 @@ def set_obj_mode(self, use_shared_mem=False):
+    ----------
+    telescope_name : str
+        Name of the telescope.
+    telescope_location : astropy EarthLocation object or MoonLocation
```
autoref

steven-murray commented on pyuvsim/utils.py:
```diff
@@ -141,7 +149,15 @@ def zenithangle_azimuth_to_altaz(zenith_angle, azimuth):
+        Extenstion to remove. If not specified, only 'uvfits', 'uvh5', 'yaml' extensions
+        are removed.
+    """
     if '.' not in filepath:
```
this function seems brittle -- thought of using the `pathlib.Path` module for this kind of stuff?

bhazelton replied:
Looks like this is only called in this module in the `check_file_exists_and_increment` function. My first reaction is I agree and I'd favor using this instead:
https://docs.python.org/3/library/os.path.html#os.path.splitext

But then I started worrying about what HERA does where it adds lots of "extensions" to filenames and I wonder if that was the concern.

@aelanman can you comment?

bhazelton replied:
I'll make an issue to follow up on this later.

steven-murray commented on pyuvsim/utils.py:
```diff
@@ -314,8 +341,9 @@ def iter_array_split(part_index, N, M):
     """
-    Estimate the memory footprint of a SkyModel by summing the sizes
-    of the data types that go into SkyModel.
+    Estimate the memory footprint of a SkyModel.
```
autoref

steven-murray commented on pyuvsim/uvsim.py:
```diff
@@ -2,6 +2,12 @@
+"""
+The primary machinery of pyuvsim.
+
+The UVTask and UVEngine classes and the functions that actually run the simulation.
```
autorefs

steven-murray commented on pyuvsim/uvsim.py:
```diff
@@ -25,11 +31,55 @@
+    ---------
+    sources : :class:~`simsetup.SkyModelData`
+        The sources to include in the visibility.
+    time : astropy Time object or float
```
autoref

steven-murray commented on pyuvsim/uvsim.py:
```diff
@@ -25,11 +31,55 @@
+    time : astropy Time object or float
+        Time at which to calculate the visibility, either an astropy Time object or a
+        float in Julian Day (will be converted to a Time object using `format='jd'`).
+    freq : astropy Quantity or float
```
autoref

steven-murray commented on pyuvsim/uvsim.py:
```diff
@@ -25,11 +31,55 @@
+    ----------
+    sources : :class:~`simsetup.SkyModelData`
+        The sources to include in the visibility.
+    time : astropy Time object
```
autoref

steven-murray commented on pyuvsim/uvsim.py:
```diff
@@ -25,11 +31,55 @@
+        The sources to include in the visibility.
+    time : astropy Time object
+        Time at which to calculate the visibility.
+    freq : astropy Quantity
```
autoref

steven-murray commented on pyuvsim/uvsim.py:
```diff
@@ -74,15 +151,84 @@ def __ge__(self, other):
+
+    Parameters
+    ----------
+    task : :class:UVTask
```
missing backticks in autoref

steven-murray commented on pyuvsim/uvsim.py:
```diff
@@ -74,15 +151,84 @@ def __ge__(self, other):
+        or beam pair changes).
+    sources : :class:~`simsetup.SkyModelData`
+        The sources to include in the visibility.
+    current_time : astropy Time object
```
autoref

steven-murray commented on pyuvsim/uvsim.py:
```diff
@@ -74,15 +151,84 @@ def __ge__(self, other):
+        The sources to include in the visibility.
+    current_time : astropy Time object
+        Time for the current calculation.
+    current_freq : astropy Quantity
```
autoref

steven-murray commented on pyuvsim/uvsim.py:
```diff
@@ -100,6 +246,15 @@ def __init__(self, task=None, update_positions=True, update_beams=True, reuse_sp
+
+        Parameters
+        ----------
+        task : :class:UVTask
```
backticks

steven-murray commented on setup.cfg:
```diff
@@ -7,6 +7,11 @@ addopts = --ignore=scripts
 # it's recommended to have max-complexity ~ 18
 max-complexity = 30
-select = B,C,E,W,T4,B9,F
+select = B,C,E,W,T4,B9,F,D
```
Instead of "selecting" specific rules, I tend to just *ignore* specific rules (i.e. select everything by default)

steven-murray commented on setup.cfg:
```diff
@@ -7,6 +7,11 @@ addopts = --ignore=scripts
 [flake8]
 ignore = W503, E203, N806
```
For clarity, I like to put each ignore code on its own line, and follow it by a comment explaining what it is and why it's being ignored. Perhaps down the line we think again about ignoring it.

steven-murray reviewed (commented):
Thanks @bhazelton --- docs are looking sooo much better! I made a lot of very small comments, thought it's better to be exhaustive.

bhazelton commented on pyuvsim/analyticbeam.py:
```diff
@@ -14,18 +15,25 @@
 
 def diameter_to_sigma(diam, freqs):
     """
+    Find the sigma that gives a beam with similar to an Airy disk.
```
I think I meant it to be "width". Is it better with or without "width"?

steven-murray replied:
oh yes, with "width"

bhazelton commented on pyuvsim/analyticbeam.py:
```diff
@@ -35,41 +43,46 @@ def diameter_to_sigma(diam, freqs):
-            If set, this sets the reference frequency for the beam width power law.
+    Calculate Jones matrices from analytic functions.
+
+    This provides similar functionality to pyuvdata.UVBeam, but its "interp" method
```
Thanks! did that some places and forgot in others.

bhazelton commented on pyuvsim/antenna.py:
```diff
@@ -11,26 +12,39 @@
     """
 
     def __init__(self, name, number, enu_position, beam_id):
+        """Initialize this antenna."""
```
Ok, I just added this to stop it complaining at me. I can ignore the code in the config.

bhazelton commented on pyuvsim/analyticbeam.py:
```diff
@@ -128,14 +139,17 @@ def interp(self, az_array, za_array, freq_array, reuse_spline=None, spline_opts=
         Returns
         -------
-        array-like of float
+        beam_vals: array-like of float
```
Thanks! There were a lot of these I fixed but I was worried I missed some.

bhazelton commented:
I could use some more help with docstrings in the scripts, benchmarks and reference_simulations folders.

bhazelton commented:
Thanks for the detailed review on this @steven-murray! I think I addressed all your comments. I also fixed the docs for files in the scripts, benchmarking and reference_simulation folders and fixed some warning handling that was making the tests fail. Are you willing to take another look?

steven-murray commented:
Thanks @bhazelton, yup, will take another look soon

steven-murray commented:
From looking at the diff, everything looks good. I'm just having a quick peruse of the compiled docs on rtd, and here are some comments:

* The main page has a list of requirements that aren't formatted well (also under Outputs)
* Bad formatting under "Testing": 
![image](https://user-images.githubusercontent.com/1272030/150614491-bd788ad2-46ec-4a85-8eae-6ca602afe808.png)
* There's something weird about how "Generating config files from data" and "parallelization" are in the side bar, but should be sub-listed under "Running a simulation"
* Missing backticks, and missing autoref here: 
![image](https://user-images.githubusercontent.com/1272030/150614731-b646f877-7f8b-4a61-9ba5-49efd45e675b.png)
* Weird formatting: 
![image](https://user-images.githubusercontent.com/1272030/150614757-66b2748c-a1a5-46a8-a405-c235aff85bd3.png)
* More weird formatting: 
![image](https://user-images.githubusercontent.com/1272030/150614784-54c01561-a531-456b-9390-c489a8e11e75.png)
* Formatting on "Gaussian": 
![image](https://user-images.githubusercontent.com/1272030/150614826-249feb00-c0a2-4c3d-a4b7-9221906a4ec8.png)
* Formatting for parameters: 
![image](https://user-images.githubusercontent.com/1272030/150614900-7af89082-1f17-45c6-b648-c17263acd440.png)
* Would be better if possible arrangements could be in list format: 
![image](https://user-images.githubusercontent.com/1272030/150614947-c4e64758-9e10-4d52-b7ab-381b31076533.png)
* Again with the list format: 
![image](https://user-images.githubusercontent.com/1272030/150614983-00143367-cdda-4443-94d2-a96c25c7e3ad.png)
* Weird formatting from `reuse_spline` on: 
![image](https://user-images.githubusercontent.com/1272030/150615070-a5f375f4-989e-4897-a8d6-84cc55068dd4.png)
* (Ah, I just realized I think the above are auto-listed attributes. Maybe better to have a heading before starting to list them?)
* I assume "Sets" is not a valid section: 
![image](https://user-images.githubusercontent.com/1272030/150615218-772e3d47-3445-4087-b67e-981ab94c53f8.png)

I think that's it!

bhazelton commented:
@steven-murray Thank for the careful read through the RTD rendered docs!! The biggest problem was that the RTD theme had broken all instances of bulleted lists (this happened before on pyuvdata but the fix I used there no longer works -- this is a new problem, see https://github.com/readthedocs/sphinx_rtd_theme/issues/1236)

I think I fixed all the things you raised except that I cannot seem to alter the behavior for how attributes are listed. They are in the docstring with a heading, but that heading is removed automagically by Sphinx. The appearance is a little better in some non-RTD themes in terms of differentiating them from parameters, but even in those themes they do not have the `Attributes` heading. If you have suggestions on this front I'd be happy to hear them!

bhazelton commented:
@steven-murray I figured out how to make the Attributes section look like the Parameters section. Do you like this better than what we had before?

bhazelton commented:
I just realized that we're using an old version of the RTD yaml on this repo. I'm going to update to the newer version in this PR as well.

bhazelton commented:
@steven-murray ok, I think I've got everything working properly now if you want to take another look.

steven-murray commented on .readthedocs.yml:
```diff
@@ -1,13 +1,26 @@
-      path: package
+      extra_requirements:
+         - doc
+    - requirements: docs/requirements.txt
```
hmmm, I'm a bit confused here... are we using the `[doc]` extra, or installing from `docs/requirements.txt`? Aren't they mutually exclusive? I prefer the former fwiw.

bhazelton replied:
Both. The requirements for building the docs with sphinx on any machine are in the `[doc]` extra. The requirements that are specific to the RTD build and hosting are in the `docs/requirements.txt`. I thought about putting those in the `[doc]` extra but I didn't like that because they aren't really required except on RTD.

steven-murray reviewed (approved):
This all looks great, thanks for this work, @bhazelton. I am pre-approving, though I do have one comment about the rtd config. Was the fix for the bullet points essentially to set the rtd theme verison exactly to 0.5.2?

bhazelton commented:
Yes, the fix for the bullets is to pin the theme to exactly 0.5.2. Older builds from that had one cause for the bullets to fail and the newer builds have a different cause for the bullets to fail 😂
